File: CSharp/Table.cs
using System;

// Элемент таблицы имен
public class Obj {      // Тип записи таблицы имен
   public string  Name; // Ключ поиска            
   public tCat    Cat;  // Категория имени        
   public tType   Type; // Тип                    
   public int     Val;  // Значение               
   public Obj     Prev; // Указатель на пред. имя 
}

// Категории имён
   public enum tCat {
      Const, Var, Type, StProc, Module, Guard
   };

// Типы
   public enum tType {
      None, Int, Bool
   }

// Таблица имен
class Table {

private static Obj   Top;    //Указатель на вершину списка    
private static Obj   Bottom; //Указатель на конец (дно) списка
private static Obj   CurrObj;

// Инициализация таблицы
public static void Init() {
   Top = null;
}

// Добавление элемента
public static void Enter(string N, tCat C, tType T, int V) {
   Obj P = new Obj();
   P.Name = String.Copy(N);
   P.Cat = C;
   P.Type = T;
   P.Val = V;
   P.Prev = Top;
   Top = P;
}

public static void OpenScope() {
   Enter("", tCat.Guard, tType.None, 0);
   if ( Top.Prev == null )
      Bottom = Top;
}

public static void CloseScope() {
   while( Top.Cat != tCat.Guard ){
      Top = Top.Prev;
   }
   Top = Top.Prev;
}

public static Obj NewName(string Name, tCat cat) {
   Obj obj = Top;

   while( obj.Cat != tCat.Guard && obj.Name != Name )
      obj = obj.Prev;
   if ( obj.Cat == tCat.Guard ) {
      obj = new Obj();
      obj.Name = String.Copy(Name);
      obj.Cat = cat;
      obj.Val = 0;
      obj.Prev = Top;
      Top = obj;
      }
   else
      Error.Message("Повторное объявление имени");
   return obj;
}

public static Obj Find(string Name) {
   Obj obj;

   Bottom.Name = String.Copy(Name);
   for( obj=Top; obj.Name != Name; obj=obj.Prev );
   if( obj == Bottom )
      Error.Message("Необъявленное имя");
   return obj;
}

public static Obj FirstVar() {
   CurrObj = Top;
   return NextVar();
}

public static Obj NextVar() {
   Obj VRef;

   while( CurrObj != Bottom && CurrObj.Cat != tCat.Var )
      CurrObj = CurrObj.Prev;
   if( CurrObj == Bottom )
      return null;
   else {
      VRef = CurrObj;
      CurrObj = CurrObj.Prev;
      return VRef;
   }
}

}

File: Java/Table.java
// Элемент таблицы имен
class Obj {       // Тип записи таблицы имен
   String Name;   // Ключ поиска            
   int Cat;       // Категория имени        
   int Typ;       // Тип                    
   int Val;       // Значение               
   Obj Prev;      // Указатель на пред. имя 
}

// Таблица имен
class Table {

// Категории имён
   static final int 
      catConst  = 1, catVar    = 2, 
      catType   = 3, catStProc = 4,
      catModule = 5, catGuard  = 6;

// Типы
   static final int 
      typNone = 0, typInt = 1, typBool = 2;

private static Obj Top;    //Указатель на вершину списка    
private static Obj Bottom; //Указатель на конец списка
private static Obj CurrObj;

// Инициализация таблицы
static void Init() {
   Top = null;
}

// Добавление элемента
static void Enter(String N, int C, int T, int V) {
   Obj P = new Obj();
   P.Name = new String(N);
   P.Cat = C;
   P.Typ = T;
   P.Val = V;
   P.Prev = Top;
   Top = P;
}

static void OpenScope() {
   Enter("", catGuard, typNone, 0);
   if ( Top.Prev == null )
      Bottom = Top;
}

static void CloseScope() {
   while( Top.Cat != catGuard ){
      Top = Top.Prev;
   }
   Top = Top.Prev;
}

static Obj NewName(String Name, int Cat) {
   Obj obj = Top;
   while(
      obj.Cat != catGuard && 
      obj.Name.compareTo(Name) != 0 
   )
      obj = obj.Prev;
   if ( obj.Cat == catGuard ) {
      obj = new Obj();
      obj.Name = new String(Name);
      obj.Cat = Cat;
      obj.Val = 0;
      obj.Prev = Top;
      Top = obj;
      }
   else
      Error.Message("Повторное объявление имени");
   return obj;
}

static Obj Find(String Name) {
   Obj obj;

   Bottom.Name = new String(Name);
   for( obj=Top; obj.Name.compareTo(Name)!=0; obj=obj.Prev );
   if( obj == Bottom )
      Error.Message("Необъявленное имя");
   return obj;
}

static Obj FirstVar() {
   CurrObj = Top;
   return NextVar();
}

static Obj NextVar() {
   Obj VRef;

   while( CurrObj != Bottom && CurrObj.Cat != catVar )
      CurrObj = CurrObj.Prev;
   if( CurrObj == Bottom )
      return null;
   else {
      VRef = CurrObj;
      CurrObj = CurrObj.Prev;
      return VRef;
   }
}

}
